
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="KMCP - K-mer-based Metagenomic Classification and Profiling">
      
      
        <meta name="author" content="Wei Shen">
      
      
        <link rel="canonical" href="https://bioinf.shenwei.me/kmcp/database/">
      
      <link rel="icon" href="../favicon.svg">
      <meta name="generator" content="mkdocs-1.2.3, mkdocs-material-8.1.5">
    
    
      
        <title>Databases - KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.ee0f47ba.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.e6a45f82.min.css">
        
          
          
          <meta name="theme-color" content="#009485">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="teal" data-md-color-accent="blue-grey">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#database" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching" class="md-header__button md-logo" aria-label="KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching" data-md-component="logo">
      
  <img src="../favicon.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Databases
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/shenwei356/kmcp/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching" class="md-nav__button md-logo" aria-label="KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching" data-md-component="logo">
      
  <img src="../favicon.svg" alt="logo">

    </a>
    KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/shenwei356/kmcp/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../download/" class="md-nav__link">
        Download
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Databases
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Databases
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#prebuilt-databases" class="md-nav__link">
    Prebuilt databases
  </a>
  
    <nav class="md-nav" aria-label="Prebuilt databases">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a-databases-for-metagenomic-profiling" class="md-nav__link">
    A). Databases for metagenomic profiling
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#b-databases-for-genome-similarity-estimation" class="md-nav__link">
    B). Databases for genome similarity estimation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#building-databases" class="md-nav__link">
    Building databases
  </a>
  
    <nav class="md-nav" aria-label="Building databases">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gtdb" class="md-nav__link">
    GTDB
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#refseq-viral-or-fungi" class="md-nav__link">
    RefSeq viral or fungi
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genbank-viral" class="md-nav__link">
    Genbank viral
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#human-genome" class="md-nav__link">
    Human genome
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#refseq-plasmid" class="md-nav__link">
    Refseq plasmid
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#building-databases-prokaryotic-genome-collections" class="md-nav__link">
    Building databases (prokaryotic genome collections)
  </a>
  
    <nav class="md-nav" aria-label="Building databases (prokaryotic genome collections)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#humgut-30691-clusters" class="md-nav__link">
    HumGut (30,691 clusters)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#progenomes" class="md-nav__link">
    proGenomes
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#building-databases-viral-genome-collections" class="md-nav__link">
    Building databases (viral genome collections)
  </a>
  
    <nav class="md-nav" aria-label="Building databases (viral genome collections)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#mgv" class="md-nav__link">
    MGV
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#building-custom-databases" class="md-nav__link">
    Building custom databases
  </a>
  
    <nav class="md-nav" aria-label="Building custom databases">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-computing-k-mers" class="md-nav__link">
    Step 1. Computing k-mers
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-building-databases" class="md-nav__link">
    Step 2. Building databases
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#merging-gtdb-and-ncbi-taxonomy" class="md-nav__link">
    Merging GTDB and NCBI taxonomy
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        Usage
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          Tutorials
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Tutorials" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Tutorials
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../tutorial/profiling/" class="md-nav__link">
        Taxonomic profiling
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../tutorial/detecting-pathogens/" class="md-nav__link">
        Detecting specific pathogens
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../tutorial/detecting-contaminated-seqs/" class="md-nav__link">
        Detecting contaminated sequences
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../tutorial/searching/" class="md-nav__link">
        Sequence and genome searching
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          Benchmarks
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Benchmarks" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          Benchmarks
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../benchmark/profiling/" class="md-nav__link">
        Taxonomic profiling
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../benchmark/searching/" class="md-nav__link">
        Sequence and genome searching
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../faq/" class="md-nav__link">
        FAQs
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356" class="md-nav__link">
        More tools
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#prebuilt-databases" class="md-nav__link">
    Prebuilt databases
  </a>
  
    <nav class="md-nav" aria-label="Prebuilt databases">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a-databases-for-metagenomic-profiling" class="md-nav__link">
    A). Databases for metagenomic profiling
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#b-databases-for-genome-similarity-estimation" class="md-nav__link">
    B). Databases for genome similarity estimation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#building-databases" class="md-nav__link">
    Building databases
  </a>
  
    <nav class="md-nav" aria-label="Building databases">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gtdb" class="md-nav__link">
    GTDB
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#refseq-viral-or-fungi" class="md-nav__link">
    RefSeq viral or fungi
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genbank-viral" class="md-nav__link">
    Genbank viral
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#human-genome" class="md-nav__link">
    Human genome
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#refseq-plasmid" class="md-nav__link">
    Refseq plasmid
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#building-databases-prokaryotic-genome-collections" class="md-nav__link">
    Building databases (prokaryotic genome collections)
  </a>
  
    <nav class="md-nav" aria-label="Building databases (prokaryotic genome collections)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#humgut-30691-clusters" class="md-nav__link">
    HumGut (30,691 clusters)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#progenomes" class="md-nav__link">
    proGenomes
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#building-databases-viral-genome-collections" class="md-nav__link">
    Building databases (viral genome collections)
  </a>
  
    <nav class="md-nav" aria-label="Building databases (viral genome collections)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#mgv" class="md-nav__link">
    MGV
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#building-custom-databases" class="md-nav__link">
    Building custom databases
  </a>
  
    <nav class="md-nav" aria-label="Building custom databases">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-computing-k-mers" class="md-nav__link">
    Step 1. Computing k-mers
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-building-databases" class="md-nav__link">
    Step 2. Building databases
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#merging-gtdb-and-ncbi-taxonomy" class="md-nav__link">
    Merging GTDB and NCBI taxonomy
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
<a href="https://github.com/shenwei356/kmcp/edit/master/docs/database.md" title="Edit this page" class="md-content__button md-icon">
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
</a>


<h1 id="database">Database</h1>
<h2 id="prebuilt-databases">Prebuilt databases</h2>
<p>All prebuilt databases and the used reference genomes are available at:</p>
<ul>
<li><a href="https://1drv.ms/u/s!Ag89cZ8NYcqtjHwpe0ND3SUEhyrp?e=QDRbEC">OneDrive</a> for <em>global users</em>.</li>
<li><a href="https://shenwei356.cowtransfer.com/s/c7220dd5901c42">CowTransfer</a> for <em>Chinese users and global users</em>.<br>
  <strong>Please click the "kmcp" icon on the left to browse directories of the current version, and choose each individual file to download.</strong>.<br></li>
</ul>
<p style="color:Tomato;">Please check file integrity with `md5sum` after downloading the files:</p>

<pre><code>md5sum -c *.kmcp.tar.gz.md5.txt
</code></pre>
<h3 id="a-databases-for-metagenomic-profiling">A). Databases for metagenomic profiling</h3>
<p>These databases are created following <a href="#building-databases">steps below</a>.
Users can also <a href="#building-custom-databases">build custom databases</a>, it's simple and fast.</p>
<p><strong>Current version</strong>: v2023.05 (<a href="https://1drv.ms/f/s!Ag89cZ8NYcqtlgAtJo--uKPNVT4t?e=KEDFrc">OneDrive</a>, <a href="https://shenwei356.cowtransfer.com/s/1c5dfb28218748">CowTransfer</a>)</p>
<table>
<thead>
<tr>
<th align="left">kingdoms</th>
<th align="left">source</th>
<th align="left">#NCBI-species</th>
<th align="left">#assemblies</th>
<th align="left">db-parameters</th>
<th align="left">size</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Bacteria and Archaea</td>
<td align="left">GTDB r214</td>
<td align="left">34395+</td>
<td align="left">85205</td>
<td align="left">k=21, chunks=10; fpr=0.3, hashes=1</td>
<td align="left">50+49 GB*</td>
</tr>
<tr>
<td align="left">Fungi</td>
<td align="left">Refseq r217</td>
<td align="left">491</td>
<td align="left">496</td>
<td align="left">k=21, chunks=10; fpr=0.3, hashes=1</td>
<td align="left">5.5 GB</td>
</tr>
<tr>
<td align="left">Viruses</td>
<td align="left">GenBank r255</td>
<td align="left">26680</td>
<td align="left">33479</td>
<td align="left">k=21, chunks=5; fpr=0.05, hashes=1</td>
<td align="left">5.9 GB</td>
</tr>
</tbody>
</table>
<p><strong>Notes</strong>: *GTDB representative genomes were split into 2 parts to build relatively small databases
which can be filled into workstations with small RAM (around 64GB).
Users need to search reads on all these small databases and merge the results before running <code>kmcp profile</code>.</p>
<p><strong>Taxonomy information</strong>: Either NCBI taxonomy or a combination of GTDB and NCBI are available:</p>
<ul>
<li>NCBI: taxdmp_2023-05-01</li>
<li>GTDB+NCBI: GTDB r214 + NCBI taxdmp_2023-05-01</li>
</ul>
<p><strong>Hardware requirements</strong></p>
<ul>
<li>Prebuilt databases were built for computers with &gt;= 32 CPU cores
in consideration of better parallelization,
and computers should have at least 64 GB.</li>
<li>By default, <code>kmcp search</code> loads the whole database into main memory
(via <a href="https://en.wikipedia.org/wiki/Mmap">mmap</a> by default) for fast searching.
Optionally, the flag <code>--low-mem</code> can be set to avoid loading the whole database,
while it's much slower, &gt;10X slower on SSD and should be much slower on HDD disks.</li>
<li><strong>To reduce memory requirements on computers without enough memory,
users can split the reference genomes into partitions
and build a smaller database for each partition, so that the biggest
database can be loaded into RAM</strong>.
This can also <strong>accelerate searching on a computer cluster, where every node searches against a small database.
After performing database searching,
search results from all small databases can be merged with <code>kmcp merge</code>
for downstream analysis</strong>.</li>
</ul>
<h3 id="b-databases-for-genome-similarity-estimation">B). Databases for genome similarity estimation</h3>
<p>Check the <a href="/kmcp/tutorial/searching">tutorial</a>.</p>
<p>FracMinHash (Scaled MinHash), links: <a href="https://1drv.ms/f/s!Ag89cZ8NYcqtlgEtskAjVVosItE7?e=CdkwhG">OneDrive</a>, <a href="https://shenwei356.cowtransfer.com/s/21f68041334440">CowTransfer</a></p>
<table>
<thead>
<tr>
<th align="left">kingdoms</th>
<th align="left">source</th>
<th align="left">#NCBI-species</th>
<th align="left">#assemblies</th>
<th align="left">parameters</th>
<th align="left">size</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Bacteria and Archaea</td>
<td align="left">GTDB r214</td>
<td align="left">34395+</td>
<td align="left">85205</td>
<td align="left">k=31, scale=1000</td>
<td align="left">2.7 GB</td>
</tr>
<tr>
<td align="left">Fungi</td>
<td align="left">Refseq r217</td>
<td align="left">491</td>
<td align="left">496</td>
<td align="left">k=31, scale=1000</td>
<td align="left">131 MB</td>
</tr>
<tr>
<td align="left"><strong>Viruses</strong></td>
<td align="left">GenBank r255</td>
<td align="left">26680</td>
<td align="left">33479</td>
<td align="left">k=31, scale=10</td>
<td align="left">2.0 GB</td>
</tr>
</tbody>
</table>
<h2 id="building-databases">Building databases</h2>
<h3 id="gtdb">GTDB</h3>
<p>Tools:</p>
<ul>
<li><a href="https://github.com/shenwei356/brename/releases">brename</a> for batching renaming files.</li>
<li><a href="https://github.com/shenwei356/rush/releases">rush</a> for executing jobs in parallel.</li>
<li><a href="https://github.com/shenwei356/seqkit/releases">seqkit</a> for FASTA file processing.</li>
<li><a href="https://github.com/shenwei356/csvtk/releases">csvtk</a> for tsv/csv data manipulations.</li>
<li><a href="https://github.com/shenwei356/taxonkit/releases">taxonkit</a> for NCBI taxonomy data manipulations.</li>
<li><a href="/kmcp/download">kmcp</a> for metagenomic profiling.</li>
</ul>
<p>Files (https://data.ace.uq.edu.au/public/gtdb/data/releases/latest/)</p>
<ul>
<li>representative genomes: gtdb_genomes_reps.tar.gz</li>
<li>metagdata: ar53_metadata.tar.gz</li>
<li>metagdata: bac120_metadata.tar.gz</li>
<li>taxdump file: https://ftp.ncbi.nih.gov/pub/taxonomy/taxdump_archive/</li>
</ul>
<p>Uncompressing and renaming:</p>
<pre><code># uncompress
mkdir -p gtdb
tar -zxvf gtdb_genomes_reps.tar.gz -C gtdb

# rename
brename -R -p '^(\w{3}_\d{9}\.\d+).+' -r '$1.fna.gz' gtdb
</code></pre>
<p>Mapping file:</p>
<pre><code># assembly accession -&gt; full head
find gtdb/ -name "*.fna.gz" \
    | rush -k 'echo -ne "{%@(.+).fna}\t$(seqkit sort --quiet -lr {} | head -n 1 | seqkit seq -n)\n" ' \
    &gt; name.map

tar -zxvf ar53_metadata.tar.gz
tar -zxvf bac120_metadata.tar.gz

# assembly accession -&gt; taxid
cat *_metadata*.tsv \
    | csvtk cut -t -f accession,ncbi_taxid \
    | csvtk replace -t -p '^.._' \
    | csvtk grep -t -P &lt;(cut -f 1 name.map) \
    | csvtk del-header \
    &gt; taxid.map


# stats (optional)
# taxdump is a directory containing NCBI taxdump files, including names.dmp, nodes.dmp, delnodes.dmp and merged.dmp
cat taxid.map  \
    | csvtk freq -Ht -f 2 -nr \
    | taxonkit lineage -r -n -L --data-dir taxdump/ \
    | taxonkit reformat -I 1 -f '{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}' --data-dir taxdump/ \
    | csvtk add-header -t -n 'taxid,count,name,rank,superkindom,phylum,class,order,family,genus,species' \
    &gt; gtdb.taxid.map.stats.tsv


# number of unique species/strains
cat taxid.map \
    | csvtk uniq -Ht -f 2 \
    | taxonkit lineage --data-dir taxdump/ -i 2 -r -n -L \
    | csvtk freq -Ht -f 4 -nr \
    | csvtk pretty -H -t \
    | tee taxid.map.uniq.stats
species           31165
strain            4073
subspecies        104
forma specialis   59
no rank           31
isolate           26
</code></pre>
<p>Masking prophage regions and removing plasmid sequences with <a href="https://github.com/apcamargo/genomad">genomad</a> (optional):</p>
<pre><code>conda activate genomad

# https://github.com/shenwei356/cluster_files
cluster_files -p '^(.+).fna.gz$' gtdb -o gtdb.genomad

# indentify prophages and plasmids sequences
# &gt; 10 days
db=~/ws/db/genomad/genomad_db
find gtdb.genomad/ -name "*.fna.gz" \
    | rush -j 6 -v db=$db \
        'genomad end-to-end --relaxed --cleanup --threads 25 {} {/}/genomad {db}' \
        -c -C genomad.rush --eta

# mask proviruses and remove plasmid sequences
find gtdb.genomad/ -name "*.fna.gz" \
    | grep -v masked \
    | rush -j 10 -v 'p={/}/genomad/{/%}_summary/{/%}' \
        'csvtk grep -t -f coordinates -p NA {p}_virus_summary.tsv \
            | sed 1d | cut -f 1 \
            &gt; {p}.v.id; \
         csvtk grep -t -f coordinates -p NA {p}_virus_summary.tsv -v \
            | sed 1d | cut -f 1 \
            | perl -ne "/^(.+)\|.+_(\d+)_(\d+)$/; \$s=\$2-1; print qq/\$1\t\$s\t\$3\n/;" \
            &gt; {p}.v.bed; \
        cut -f 1 {p}_plasmid_summary.tsv | sed 1d &gt; {p}.p.id; \
        bedtools maskfasta -fi &lt;(seqkit grep --quiet -v -f &lt;(cat {p}.p.id {p}.v.id) {}) \
                            -bed {p}.v.bed -fo {}.masked.fna; \
        gzip -f {}.masked.fna' \
        -c -C genomad.b.rush --eta

# replace masked sequence with k-Ns, for more accurate genome size.
find gtdb.genomad/ -name "*.fna.gz.masked.fna.gz" \
    | rush -j 20 'seqkit replace -s -i -p N+ -r NNNNNNNNNNNNNNNNNNNNN {} -o {...}.masked2.fna.gz' \
        -c -C genomad.c.rush --eta

# stats
time find gtdb.genomad/ -name "*.fna.gz" \
    | seqkit stats -j 20 --infile-list - -T -b -o gtdb.genomad.stats.tsv

# create symbol links in another directory, and rename
mkdir -p gtdb_masked
cluster_files  -p '^(.+).fna.gz.masked2.fna.gz$' gtdb.genomad -o gtdb_masked
brename -R -p .masked.fna.gz$ gtdb_masked -d
</code></pre>
<p>Building database (all k-mers, for profiling on short-reads):</p>
<pre><code>input=gtdb

# compute k-mers
#   sequence containing "plasmid" in name are ignored,
#   reference genomes are split into 10 chunks with 100bp overlap
#   k = 21
kmcp compute -I $input -O gtdb-k21-n10 -k 21 -n 10 -l 150 -B plasmid \
    --log gtdb-k21-n10.log -j 32 --force

# build database
#   number of index files: 32, for server with &gt;= 32 CPU cores
#   bloom filter parameter:
#     number of hash function: 1
#     false positive rate: 0.3
kmcp index -j 32 -I gtdb-k21-n10 -O gtdb.kmcp -n 1 -f 0.3 \
    --log gtdb.kmcp.log --force

# cp taxid and name mapping file to database directory
cp taxid.map name.map gtdb.kmcp/

# clean up
rm -rf gtdb-k21-n10
</code></pre>
<p>Building small databases (all k-mers, for profiling with a computer cluster or computer with limited RAM):</p>
<pre><code>input=gtdb

find $input -name "*.fna.gz" &gt; $input.files.txt

# sort files by genome size, so we can split them into chunks with similar genome sizes
cp $input.files.txt $input.files0.txt
cat $input.files0.txt \
    | rush --eta -k 'echo -e {}"\t"$(seqkit stats -T {} | sed 1d | cut -f 5)' \
    &gt; $input.files0.size.txt
cat $input.files0.size.txt \
    | csvtk sort -Ht -k 2:nr \
    | csvtk cut -t -f 1 \
    &gt; $input.files.txt

# number of databases
n=2

# split into $n chunks using round robin distribution
split -n r/$n -d  $input.files.txt $input.n$n-

# create database for every chunks
for f in $input.n$n-*; do 
    echo $f

    # compute k-mers
    #   sequence containing "plasmid" in name are ignored,
    #   reference genomes are split into 10 chunks with 100bp overlap
    #   k = 21
    kmcp compute -i $f -O $f-k21-n10 -k 21 -n 10 -l 150 -B plasmid \
        --log $f-k21-n10.log --force

    # build database
    #   number of index files: 32, for server with &gt;= 32 CPU cores
    #   bloom filter parameter:
    #     number of hash function: 1
    #     false positive rate: 0.3
    kmcp index -j 32 -I $f-k21-n10 -O $f.kmcp -n 1 -f 0.3 \
        --log $f.kmcp.log --force

    # cp taxid and name mapping file to database directory
    cp taxid.map name.map $f.kmcp/
done

# rename small databases
# for example:
#   [INFO] checking: [ ok ] 'gtdb.n2-00.kmcp' -&gt; 'gtdb.part_1.kmcp'
#   [INFO] checking: [ ok ] 'gtdb.n2-01.kmcp' -&gt; 'gtdb.part_2.kmcp'
#   [INFO] 2 path(s) to be renamed

brename -D --only-dir -p "n$n-\d+" -r "part_{nr}" -d


# clean up
for f in $input.n$n-*; do
    rm -rf $f-k21-n10
done
</code></pre>
<h3 id="refseq-viral-or-fungi">RefSeq viral or fungi</h3>
<p>Tools</p>
<ul>
<li><a href="https://github.com/pirovc/genome_updater">genome_updater</a> (0.5.1) for downloading genomes from NCBI.</li>
</ul>
<p>Downloading viral and fungi sequences:</p>
<pre><code>name=fungi
# name=viral

# -k for dry-run
# -i for fix
time genome_updater.sh \
    -d "refseq"\
    -g $name \
    -c "" \
    -l "" \
    -f "genomic.fna.gz" \
    -o "refseq-$name" \
    -t 12 \
    -m -a -p

# cd to 2021-09-30_19-35-19

# taxdump
mkdir -p taxdump
tar -zxvf taxdump.tar.gz -C taxdump

# assembly accession -&gt; taxid
cut -f 1,6 assembly_summary.txt &gt; taxid.map    
# assembly accession -&gt; name
cut -f 1,8 assembly_summary.txt &gt; name.map


# stats (optional)
cat taxid.map  \
    | csvtk freq -Ht -f 2 -nr \
    | taxonkit lineage -r -n -L --data-dir taxdump/ \
    | taxonkit reformat -I 1 -f '{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}' --data-dir taxdump/ \
    | csvtk add-header -t -n 'taxid,count,name,rank,superkindom,phylum,class,order,family,genus,species' \
    &gt; refseq-$name.taxid.map.stats.tsv

seqkit stats -T -j 12 --infile-list &lt;(find files -name "*.fna.gz") &gt; files.stats.tsv



# ------------------------------------------------------ 
# create another directory linking to genome files

input=files.renamed

mkdir -p $input
# create soft links
cd $input
find ../files -name "*.fna.gz" \
    | rush 'ln -s {}'
cd ..    
# rename
brename -R -p '^(\w{3}_\d{9}\.\d+).+' -r '$1.fna.gz' $input
</code></pre>
<p>Building database (all k-mers, for profiling on short-reads):</p>
<pre><code># -----------------------------------------------------------------
# for viral, only splitting into 10 chunks
name=viral

input=files.renamed

# -------------
# all kmers

kmcp compute -I $input -O refseq-$name-k21-n10 \
    -k 21 --seq-name-filter plasmid \
    --split-number 10 --split-overlap 150 \
    --log refseq-$name-k21-n10.log -j 32 --force

# viral genomes are small:
#   using small false positive rate: 0.001
#   using more hash functions: 3
kmcp index -I refseq-$name-k21-n10/ -O refseq-viral.kmcp \
    -j 32 -f 0.001 -n 3 -x 100K \
    --log refseq-viral.kmcp.log --force

# cp taxid and name mapping file to database directory
cp taxid.map name.map refseq-viral.kmcp/


# -----------------------------------------------------------------
# for fungi
name=fungi

input=files.renamed


# -------------
# all kmers

kmcp compute -I $input -O refseq-$name-k21-n10 \
    -k 21 --seq-name-filter plasmid \
    --split-number 10 --split-overlap 150 \
    --log refseq-$name-k21-n10.log --force

kmcp index -I refseq-$name-k21-n10/ -O refseq-fungi.kmcp \
    -j 32 -f 0.3 -n 1 \
    --log refseq-fungi.kmcp.log --force

# cp taxid and name mapping file to database directory
cp taxid.map name.map refseq-fungi.kmcp/
</code></pre>
<h3 id="genbank-viral">Genbank viral</h3>
<p>Tools</p>
<ul>
<li><a href="https://github.com/pirovc/genome_updater">genome_updater</a> for downloading genomes from NCBI.</li>
</ul>
<p>Downloading viral sequences:</p>
<pre><code>name=viral

# -k for dry-run
# -i for fix
# keep at most 5 genomes for a taxid:
#    genome_updater v0.5.1: -A 5 -c "" -l ""
#    genome_updater v0.2.5: -j taxids:5 -c "all" -l "all"
time genome_updater.sh \
    -d "genbank"\
    -A 5 \
    -g $name \
    -c "" \
    -l "" \
    -f "genomic.fna.gz" \
    -o "genbank-$name" \
    -t 12 \
    -m -a -p

# cd genbank-viral/2021-12-06_15-27-37/

# taxdump
mkdir -p taxdump
tar -zxvf taxdump.tar.gz -C taxdump

# assembly accession -&gt; taxid
cut -f 1,6 assembly_summary.txt &gt; taxid.map    
# assembly accession -&gt; name
cut -f 1,8 assembly_summary.txt &gt; name.map


# stats (optional)
cat taxid.map  \
    | csvtk freq -Ht -f 2 -nr \
    | taxonkit lineage -r -n -L --data-dir taxdump/ \
    | taxonkit reformat -I 1 -f '{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}' --data-dir taxdump/ \
    | csvtk add-header -t -n 'taxid,count,name,rank,superkindom,phylum,class,order,family,genus,species' \
    &gt; genbank-viral.taxid.map.stats.tsv

seqkit stats -T -j 12 --infile-list &lt;(find files -name "*.fna.gz") &gt; files.stats.tsv



# ------------------------------------------------------ 
# create another directory linking to genome files

input=files.renamed

mkdir -p $input
# create soft links
cd $input
find ../files -name "*.fna.gz" \
    | rush 'ln -s {}'
cd ..    
# rename
brename -R -p '^(\w{3}_\d{9}\.\d+).+' -r '$1.fna.gz' $input
</code></pre>
<p>Keep at most 5 genomes for a taxid (optional)</p>
<pre><code># -----------------------------------------------------------------
# keep at most 5 genomes for a taxid

# backup
cat taxid.map &gt; taxid.all.map
cat name.map &gt; name.all.map

# keep at most 5 ids for a taxid
cat taxid.all.map \
    | csvtk sort -Ht -k 2:n -k 1:n \
    | csvtk uniq -Ht -f 2 -n 5 \
    &gt; taxid.map

cat name.all.map \
    | csvtk grep -Ht -P &lt;(cut -f 1 taxid.map) \
    &gt; name.map

# organize files
input=files.renamed
output=files.renamed.slim
mkdir -p $output 
cd $output
find ../$input -name "*.fna.gz" \
    | csvtk mutate -Ht -p '/([^\/]+).fna.gz' \
    | csvtk grep -Ht -f 2 -P &lt;(cut -f 1 ../taxid.map) \
    | rush 'ln -s {1}'
cd ..

# check number of genomes
ls $output | wc -l
wc -l taxid.map
</code></pre>
<p>Building database (all k-mers, for profiling on short-reads):</p>
<pre><code># -----------------------------------------------------------------
name=viral

# input=files.renamed.slim
input=files.renamed


# ----------------
# all kmers

kmcp compute -I $input -O genbank-$name-k21-n10 \
    -k 21 --seq-name-filter plasmid \
    --split-number 10 --split-overlap 150 \
    --log genbank-$name-k21-n10.log --force

# viral genomes are small:
#   using small false positive rate: 0.05
#   still using one hash function: 1
kmcp index -I genbank-$name-k21-n10/ -O genbank-viral.kmcp \
    -j 32 -f 0.05 -n 1 -x 100K -8 1M \
    --log genbank-viral.kmcp.log --force

# cp taxid and name mapping file to database directory
cp taxid.map name.map genbank-viral.kmcp/
</code></pre>
<p>Building small databases (all k-mers, for profiling with a computer cluster or computer with limited RAM):</p>
<pre><code>name=genbank-viral

input=files.renamed.slim

find $input -name "*.fna.gz" &gt; $input.files.txt

# sort files by genome size, so we can split them into chunks with similar genome sizes
cp $input.files.txt $input.files0.txt
cat $input.files0.txt \
    | rush -k 'echo -e {}"\t"$(seqkit stats -T {} | sed 1d | cut -f 5)' \
    &gt; $input.files0.size.txt
cat $input.files0.size.txt \
    | csvtk sort -Ht -k 2:nr \
    | csvtk cut -t -f 1 \
    &gt; $input.files.txt

# number of databases
n=4

# split into $n chunks using round robin distribution
split -n r/$n -d  $input.files.txt $name.n$n-

# create database for every chunks
for f in $name.n$n-*; do 
    echo $f

    kmcp compute -i $f -O $f-k21-n10 \
        -k 21 --seq-name-filter plasmid \
        --split-number 10 --split-overlap 150 \
        --log $f-k21-n10.log -j 24 --force

    # viral genomes are small:
    #   using small false positive rate: 0.001
    #   using more hash functions: 3
    kmcp index -I $f-k21-n10/ -O $f.kmcp \
        -j 24 -f 0.05 -n 1 -x 100K -8 1M \
        --log $f.kmcp.log --force

    # cp taxid and name mapping file to database directory
    cp taxid.map name.map $f.kmcp/
done
</code></pre>
<h3 id="human-genome">Human genome</h3>
<p>Downloading human genome file from <a href="https://github.com/marbl/CHM13">CHM13</a>:</p>
<pre><code># v1.1: wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/009/914/755/GCA_009914755.3_CHM13_T2T_v1.1/GCA_009914755.3_CHM13_T2T_v1.1_genomic.fna.gz

# v2.0
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/009/914/755/GCA_009914755.4_T2T-CHM13v2.0/GCA_009914755.4_T2T-CHM13v2.0_genomic.fna.gz
</code></pre>
<p>Building database (all k-mers, &lt; 10min):</p>
<pre><code># v1.1: input=GCA_009914755.3_CHM13_T2T_v1.1_genomic.fna.gz

# v2.0
input=GCA_009914755.4_T2T-CHM13v2.0_genomic.fna.gz

# splitting human genome into 3072 chunks, each with ~1Mb
# The regular expression '^(\w{3}_\d{9}\.\d+).+' is for extracting 'GCA_009914755.4' from the file name.
kmcp compute $input -O human-chm13-k21-n3072 \
    --ref-name-regexp '^(\w{3}_\d{9}\.\d+).+' \
    -k 21 \
    --split-number 3072 --split-overlap 150 \
    --log human-chm13-k21-n3072.log -j 32 --force

# using a small false positive rate: 0.05
# using one hash function: 1
kmcp index -I human-chm13-k21-n3072/ -O human-chm13.kmcp \
    -j 8 -f 0.05 -n 1 \
    --log human-chm13.kmcp.log --force

# taxid.map
echo -ne "GCA_009914755.4\t9606\n" &gt; taxid.map

# name.map
echo -ne "GCA_009914755.4\tHomo sapiens isolate CHM13\n" &gt; name.map

# cp name mapping file to database directory
cp taxid.map name.map human-chm13.kmcp/

# database size
$ du -sh human-chm13.kmcp
6.7G    human-chm13.kmcp
</code></pre>
<h3 id="refseq-plasmid">Refseq plasmid</h3>
<p>Downloading plasmid sequences:</p>
<pre><code>wget https://ftp.ncbi.nlm.nih.gov/refseq/release/plasmid/
cat index.html \
    | perl -ne 'next unless /"(plasmid.*genomic.fna.gz)"/; print "$1\n"' &gt; files.txt

baseurl=https://ftp.ncbi.nlm.nih.gov/refseq/release/plasmid/
outdir=archives; mkdir -p $outdir
cat files.txt \
    | rush -j 12 -v b=$baseurl -v out=$outdir \
        'wget -c {b}/{} -o /dev/null -O {out}/{}' -c -C download.rush


# name mapping
seqkit seq -n archives/*.fna.gz \
    | sed -E 's/\s+/\t/' \
    &gt; name.map

# length stats
seqkit stats -b -a -j 12 -T archives/*.fna.gz &gt; archives.stats.tsv


# split to individual files
for f in archives/*.fna.gz; do \
    seqkit split2 -s 1 --quiet $f -O refseq-plasmid; \
done
# rename files with sequence ID
find refseq-plasmid -name "*.fna.gz" \
    | rush 'mv {} {/}/$(seqkit seq -ni {}).fna.gz'
</code></pre>
<p>Building database (all k-mers):</p>
<pre><code>name=plasmid

input=refseq-plasmid

kmcp compute -I $input -O refseq-$name-k21-n10 \
    -k 21 --circular \
    --split-number 10 --split-overlap 150 \
    --log refseq-$name-k21-n10.log -j 32 --force

#   using small false positive rate: 0.01
#   using more hash functions: 3
kmcp index -I refseq-$name-k21-n10/ -O refseq-$name.kmcp \
    -j 32 -f 0.01 -n 3 -x 200K -X 1024 \
    --log refseq-$name.kmcp.log --force

# cp name mapping file to database directory
cp name.map refseq-$name.kmcp/
</code></pre>
<p>Building database (FracMinHash/Scaled MinHash):</p>
<pre><code>name=plasmid

input=refseq-plasmid

kmcp compute -I $input -O refseq-$name-k31-D10 \
    -k 31 --circular --scale 10 \
    --log refseq-$name-k31-D10.log -j 32 --force

#   using small false positive rate: 0.01
#   using more hash functions: 3
kmcp index -I refseq-$name-k31-D10/ -O refseq-$name.minhash.kmcp \
    -j 8 -f 0.01 -n 3 -x 100K -X 1024 \
    --log refseq-$name.minhash.kmcp.log --force

# cp name mapping file to database directory
cp name.map refseq-$name.minhash.kmcp/
</code></pre>
<p>Building database (Closed Syncmer):</p>
<pre><code>name=plasmid

input=refseq-plasmid

kmcp compute -I $input -O refseq-$name-k31-S21 \
    -k 31 --circular --syncmer-s 21 \
    --log refseq-$name-k31-S21.log -j 32 --force

#   using small false positive rate: 0.01
#   using more hash functions: 3
kmcp index -I refseq-$name-k31-S21/ -O refseq-$name.syncmer.kmcp \
    -j 8 -f 0.01 -n 3 -x 100K -X 1024 \
    --log refseq-$name.syncmer.kmcp.log --force

# cp name mapping file to database directory
cp name.map refseq-$name.syncmer.kmcp/
</code></pre>
<h2 id="building-databases-prokaryotic-genome-collections">Building databases (prokaryotic genome collections)</h2>
<h3 id="humgut-30691-clusters">HumGut (30,691 clusters)</h3>
<p><a href="http://arken.nmbu.no/~larssn/humgut/">HumGut</a> is a comprehensive Human Gut prokaryotic genomes collection filtered by metagenome data.</p>
<blockquote>
<p>In this work, we aimed to create a collection of the most prevalent healthy human gut prokaryotic genomes,
 to be used as a reference database, including both MAGs from the human gut and ordinary RefSeq genomes.</p>
</blockquote>
<p>Dataset</p>
<ul>
<li>Genomes: <a href="http://arken.nmbu.no/~larssn/humgut/HumGut.tar.gz">HumGut.tar.gz</a></li>
<li>Metadata: <a href="http://arken.nmbu.no/~larssn/humgut/HumGut.tsv">HumGut.tsv</a></li>
<li>Taxdump files:<ul>
<li>NCBI taxonomy: <a href="http://arken.nmbu.no/~larssn/humgut/ncbi_names.dmp">ncbi_names.dump</a>
  and <a href="http://arken.nmbu.no/~larssn/humgut/ncbi_names.dmp">ncbi_nodes.dump</a></li>
<li>GTDB taxomomy: <a href="http://arken.nmbu.no/~larssn/humgut/gtdb_names.dmp">gtdb_names.dump</a>
  and <a href="http://arken.nmbu.no/~larssn/humgut/gtdb_names.dmp">gtdb_nodes.dump</a></li>
</ul>
</li>
</ul>
<p>Uncompressing and renaming:</p>
<pre><code># uncompress
tar -zxvf HumGut.tar.gz

# taxdump files
mkdir taxdump
mv ncbi_names.dmp taxdump/names.dmp
mv ncbi_nodes.dmp taxdump/nodes.dmp
</code></pre>
<p>Mapping file:</p>
<pre><code># assembly accession -&gt; taxid
cat HumGut.tsv \
    | csvtk cut -t -f genome_file,ncbi_tax_id \
    | csvtk replace -f genome_file -t -p '\.fna\.gz$' \
    | csvtk del-header \
    &gt; taxid.map

# assembly accession -&gt; name
cat HumGut.tsv \
    | csvtk cut -t -f genome_file,ncbi_organism_name \
    | csvtk replace -f genome_file -t -p '\.fna\.gz$' \
    | csvtk del-header \
    &gt; name.map
</code></pre>
<p>Stats:</p>
<pre><code># -------------------------------------------------------------
# NCBI taxonomy

# number of species/strains
cat taxid.map \
    | taxonkit lineage --data-dir taxdump-ncbi -i 2 -r -n -L \
    | csvtk freq -Ht -f 4 -nr \
    | csvtk pretty -H -t \
    | tee taxid.map.stats
species      23604
strain       6816
subspecies   161
no rank      69
serotype     38
genus        3

# number of unique species/strains
cat taxid.map \
    | csvtk uniq -Ht -f 2 \
    | taxonkit lineage --data-dir taxdump-ncbi -i 2 -r -n -L \
    | csvtk freq -Ht -f 4 -nr \
    | csvtk pretty -H -t \
    | tee taxid.map.uniq.stats;
species      1240
strain       458
subspecies   15
genus        1
no rank      1
serotype     1

# -------------------------------------------------------------
# GTDB taxonomy

cat HumGut.tsv \
    | csvtk cut -t -f genome_file,gtdbtk_tax_id \
    | csvtk replace -f genome_file -t -p '\.fna\.gz$' \
    | csvtk del-header \
    &gt; taxid-gtdb.map

cat taxid-gtdb.map \
    | csvtk uniq -Ht -f 2 \
    | taxonkit lineage --data-dir taxdump-gtdb -i 2 -r -n -L \
    | csvtk freq -Ht -f 4 -nr \
    | csvtk pretty -H -t \
    | tee taxid-gtdb.map.uniq.stats
species   2810
genus     434
family    52
order     12
class     2
</code></pre>
<p>Building database:</p>
<pre><code># compute k-mers
#   sequence containing "plasmid" in name are ignored,
#   reference genomes are split into 10 chunks
#   k = 21
kmcp compute -I fna/ -k 21 -n 10 -B plasmid -O humgut-k21-n10 -j 32 --force

# build database
#   number of index files: 32, for server with &gt;= 32 CPU cores
#   bloom filter parameter:
#     number of hash function: 1
#     false positive rate: 0.3
kmcp index -j 32 -I humgut-k21-n10 -O humgut.kmcp -n 1 -f 0.3

cp taxid.map taxid-gtdb.map humgut.kmcp/
</code></pre>
<h3 id="progenomes">proGenomes</h3>
<p><a href="https://progenomes.embl.de/">proGenomes</a> v2.1 provides 84,096 consistently annotated bacterial
and archaeal genomes from over 12000 species.</p>
<p>Dataset:</p>
<ul>
<li>Representative genomes: <a href="https://progenomes.embl.de/data/repGenomes/freeze12.contigs.representatives.fasta.gz">contigs.representatives.fasta.gz</a></li>
<li>NCBI taxonomy: <a href="https://progenomes.embl.de/data/proGenomes2.1_specI_lineageNCBI.tab">proGenomes2.1_specI_lineageNCBI.tab</a></li>
</ul>
<p>Organize sequences:</p>
<pre><code># download
wget https://progenomes.embl.de/data/repGenomes/freeze12.contigs.representatives.fasta.gz

# unzip for splitting by genome ID
time seqkit seq freeze12.contigs.representatives.fasta.gz -o freeze12.contigs.representatives.fasta

# split by genome ID
seqkit split --by-id --two-pass --id-regexp '(^\d+\.\w+)\.' freeze12.contigs.representatives.fasta --out-dir genomes

# batch rename
brename -p '^.+id_' genomes

# compress genomes for saving space
find genomes/ -name "*.fasta" | rush 'gzip {}'
</code></pre>
<p>Mapping file:</p>
<pre><code># download
wget https://progenomes.embl.de/data/proGenomes2.1_specI_lineageNCBI.tab

ls genomes/ | sed 's/.fasta.gz//' &gt; id.txt

# id -&gt; taxid
cut -f 1 proGenomes2.1_specI_lineageNCBI.tab \
    | awk -F . '{print $0"\t"$1}' \
    | csvtk grep -Ht -P id.txt \
    &gt; taxid.map

cat taxid.map \
    | csvtk uniq -Ht -f 2 \
    | taxonkit lineage --data-dir taxdump -i 2 -r -n -L \
    | csvtk freq -Ht -f 4 -nr \
    | csvtk pretty -H -t \
    | tee taxid.map.uniq.stats
species           8088
strain            3262
subspecies        37
isolate           25
no rank           16
forma specialis   14
biotype           1


# use the taxonomy verion of the refseq-virus: 2021-10-01
# id -&gt; name
cat taxid.map \
    | taxonkit lineage --data-dir taxdump -i 2 -n -L \
    | cut -f 1,3 \
    &gt; name.map
</code></pre>
<p>Building database:</p>
<pre><code># compute k-mers
#   sequence containing "plasmid" in name are ignored,
#   reference genomes are split into 10 chunks
#   k = 21
kmcp compute -I genomes/ -k 21 -n 10 -B plasmid -O progenomes-k21-n10 --force

# build database
#   number of index files: 32, for server with &gt;= 32 CPU cores
#   bloom filter parameter:
#     number of hash function: 1
#     false positive rate: 0.3
kmcp index -j 32 -I progenomes-k21-n10 -O progenomes.kmcp -n 1 -f 0.3

cp taxid.map name.map progenomes.kmcp/
</code></pre>
<h2 id="building-databases-viral-genome-collections">Building databases (viral genome collections)</h2>
<h3 id="mgv">MGV</h3>
<blockquote>
<p>Bacteriophages have important roles in the ecology of the human gut microbiome but are under-represented in reference data-
bases. To address this problem, we assembled the Metagenomic Gut Virus catalogue that comprises 189,680 viral genomes
from 11,810 publicly available human stool metagenomes. Over 75% of genomes represent double-stranded DNA phages that
infect members of the Bacteroidia and Clostridia classes. Based on sequence clustering we identified 54,118 candidate viral spe-
cies, 92% of which were not found in existing databases. The Metagenomic Gut Virus catalogue improves detection of viruses
in stool metagenomes and accounts for nearly 40% of CRISPR spacers found in human gut Bacteria and Archaea. We also pro-
duced a catalogue of 459,375 viral protein clusters to explore the functional potential of the gut virome. This revealed tens of
thousands of diversity-generating retroelements, which use error-prone reverse transcription to mutate target genes and may
be involved in the molecular arms race between phages and their bacterial hosts.</p>
<p>https://doi.org/10.1038/s41564-021-00928-6
https://portal.nersc.gov/MGV/</p>
</blockquote>
<p>Basic information (optional)</p>
<pre><code>$ seqkit  stats mgv_contigs.fna 
file             format  type  num_seqs        sum_len  min_len   avg_len  max_len
mgv_contigs.fna  FASTA   DNA    189,680  8,803,222,510    1,244  46,410.9  553,716

# Genome completeness
Complete: n=26,030
&gt;90% complete: n=53,220
50-90% complete: n=110,430

$ seqkit  seq -n mgv_contigs.fna | head -n 3
MGV-GENOME-0364295
MGV-GENOME-0364296
MGV-GENOME-0364303
</code></pre>
<p>Stats (optional)</p>
<pre><code>cat mgv_contig_info.tsv \
    | csvtk cut -t -f completeness \
    | csvtk plot hist -o completeness.hist.png

# Complete
$ cat mgv_contig_info.tsv \
    | csvtk filter2 -t -f '$completeness == 100' \
    | csvtk nrows 
32577

# &gt;90% complete
$ cat mgv_contig_info.tsv \
    | csvtk filter2 -t -f '$completeness &gt;= 90' \
    | csvtk nrows 
78814  # &lt; 79250

# checkv_quality
$ cat mgv_contig_info.tsv \
    | csvtk cut -t -f checkv_quality \
    | csvtk freq -t -nr | more
checkv_quality  frequency
Medium-quality  110430
High-quality    53220
Complete        26030

# &gt;90% complete &amp;&amp; checkv_quality == High-quality/Complete
$ cat mgv_contig_info.tsv \
    | csvtk filter2 -t -f '$completeness &gt;= 90 &amp;&amp; $checkv_quality != "Medium-quality"' \
    | csvtk nrows 
78813
</code></pre>
<p>Stats of high-quality genomes (optional)</p>
<pre><code># high-quality genomes
cat mgv_contig_info.tsv \
    | csvtk filter2 -t -f '$completeness &gt;= 90 &amp;&amp; $checkv_quality != "Medium-quality"' \
    &gt; mgv.hq.tsv

# number of families
cat mgv.hq.tsv \
    | csvtk freq -t -f ictv_family -nr \
    | csvtk nrow -t
19

# number of species
cat mgv.hq.tsv \
    | csvtk freq -t -f votu_id -nr \
    | csvtk nrow -t
26779

# baltimore
cat mgv.hq.tsv \
    | csvtk freq -t -f baltimore -nr \
    | csvtk pretty -t
baltimore   frequency
---------   ---------
dsDNA       76527
ssDNA       2215
NULL        62
DNA         7
dsDNA-RT    1
ssRNA-RT    1

# prophage?
cat mgv.hq.tsv \
    | csvtk freq -t -f prophage -nr \
    | csvtk pretty -t
prophage   frequency
--------   ---------
No         58366
Yes        20447

# species with both prophage and lytic phages
cat mgv.hq.tsv \
    | csvtk freq -t -f votu_id,prophage \
    | csvtk freq -t -f votu_id \
    | csvtk filter2 -t -f '$frequency &gt; 1' \
    | csvtk nrow -t
2745
</code></pre>
<p>Prepare genomes:</p>
<pre><code># ID
cat mgv_contig_info.tsv \
    | csvtk filter2 -t -f '$completeness &gt;= 90 &amp;&amp; $checkv_quality != "Medium-quality"' \
    | csvtk cut -t -f contig_id \
    | csvtk del-header \
    &gt; mgv.hq.tsv.id

# extract sequences
seqkit grep -f mgv.hq.tsv.id mgv_contigs.fna -o mgv.hq.fasta.gz

# split into files with one genome
seqkit split2 -s 1 mgv.hq.fasta.gz -O mgv

# rename with 
find mgv/ -name "*.fasta.gz" \
    | rush -j 20 'mv {} {/}/$(seqkit seq -ni {}).fa.gz'
</code></pre>
<p>Create taxdump files and <code>taxid.map</code> with taxonkit (version &gt;= v0.12.0):</p>
<pre><code>cat mgv_contig_info.tsv \
        | csvtk cut -t -f ictv_order,ictv_family,ictv_genus,votu_id,contig_id \
    | csvtk del-header \
    &gt; mgv.taxonomy.tsv

taxonkit create-taxdump mgv.taxonomy.tsv --out-dir mgv-taxdump \
    --force -A 5 -R order,family,genus,species

cp mgv-taxdump/taxid.map .

# name.map
cat mgv-taxdump/taxid.map \
    | taxonkit lineage --data-dir mgv-taxdump/ -i 2 \
    | cut -f 1,3 \
    &gt; name.map

head -n 5 name.map 
MGV-GENOME-0364295      Caudovirales;crAss-phage;OTU-61123
MGV-GENOME-0364296      Caudovirales;crAss-phage;OTU-61123
MGV-GENOME-0364303      Caudovirales;crAss-phage;OTU-05782
MGV-GENOME-0364311      Caudovirales;crAss-phage;OTU-01114
MGV-GENOME-0364312      Caudovirales;crAss-phage;OTU-23935
</code></pre>
<p>Building database:</p>
<pre><code># compute k-mers
#   reference genomes are split into 10 chunks
#   k = 21
kmcp compute -I mgv/ -k 21 -n 10 -l 150 -O mgv-k21-n10 --force

# viral genomes are small:
#   using small false positive rate: 0.05
#   still using one hash function: 1
kmcp index -j 32 -I mgv-k21-n10 -O mgv.kmcp \
    -j 32 -f 0.05 -n 1 -x 100K -8 1M \
    --log mgv.kmcp.log

cp taxid.map name.map mgv.kmcp/
</code></pre>
<h2 id="building-custom-databases">Building custom databases</h2>
<p>Files:</p>
<ol>
<li>Genome files<ul>
<li>(Gzip-compressed) FASTA/Q format.</li>
<li>One genome per file <strong>with the reference identifier in the file name</strong>.</li>
</ul>
</li>
<li>TaxId mapping file (for metagenomic profiling)<ul>
<li>Two-column (reference identifier and TaxId) tab-delimited.</li>
</ul>
</li>
<li><a href="ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz">NCBI taxonomy dump files</a> (for metagenomic profiling).
  You can use <a href="https://bioinf.shenwei.me/taxonkit/usage/#create-taxdump">taxonkit create-taxdump</a>
  to create NCBI-style taxdump files for custom genome collections, which also generates a TaxId mapping file.<ul>
<li><code>names.dmp</code></li>
<li><code>nodes.dmp</code></li>
<li><code>merged.dmp</code> (optional)</li>
<li><code>delnodes.dmp</code> (optional)</li>
</ul>
</li>
</ol>
<p>Tools:</p>
<ul>
<li><a href="/kmcp/download">kmcp</a> for metagenomic profiling.</li>
<li><a href="https://github.com/shenwei356/rush/releases">rush</a> for executing jobs in parallel.</li>
<li><a href="https://github.com/shenwei356/brename/releases">brename</a> for batching renaming files (optional)</li>
</ul>
<p><strong>Memory notes</strong>:</p>
<p>By default, <code>kmcp search</code> loads the whole database into main memory (RAM) for fast searching.
Optionally, the flag <code>--low-mem</code> can be set to avoid loading the whole database,
while it's much slower, &gt;10X slower on SSD and should be much slower on HDD disks.</p>
<p>Another way is dividing the reference genomes into several parts
and building smaller databases for all parts, so that the biggest
database can be loaded into RAM. After performing database searching,
search results on all small databases can be merged with <code>kmcp merge</code>
for downstream analysis.</p>
<p>Buiding small databases can also accelerate searching on a <em>computer cluster</em>,
where every node searches a part of the database.</p>
<h3 id="step-1-computing-k-mers">Step 1. Computing k-mers</h3>
<p><strong>Input:</strong></p>
<ol>
<li>Input plain or gzipped FASTA/Q files can be given via positional arguments or
    the flag <code>-i/--infile-list</code> with the list of input files,</li>
<li>Or a directory containing sequence files via the flag <code>-I/--in-dir</code>,
    with multiple-level sub-directories allowed. A regular expression
    for matching sequencing files is available via the flag <code>-r/--file-regexp</code>.
    The default pattern matches files with extension of 
    <code>fa</code>, <code>fasta</code>, <code>fna</code>, <code>fq</code>, <code>fastq</code>, <code>fa.gz</code>, <code>fasta.gz</code>,
    <code>fna.gz</code>, <code>fq.gz</code> or <code>fastq.gz</code>.</li>
</ol>
<p>You may rename the sequence files for convenience using <a href="https://github.com/shenwei356/brename">brename</a>.
because the sequence/genome identifier in the index and search results would be:</p>
<ol>
<li>
<p>For the default mode (computing k-mers for the whole file):</p>
<pre><code>the basename of file with common FASTA/Q file extension removed,
captured via the flag -N/--ref-name-regexp.
</code></pre>
</li>
<li>
<p>For splitting sequence mode (see details below):</p>
<pre><code>same to 1).
</code></pre>
</li>
<li>
<p>For computing k-mers for each sequence:</p>
<pre><code>the sequence identifier.
</code></pre>
</li>
</ol>
<p>Unwanted sequences like plasmid sequences can be <strong>filtered out</strong> by
the name via regular expression(s) (<code>-B/--seq-name-filter</code>).</p>
<p><strong>How to compute k-mers</strong>:
By default, <code>kmcp</code> computes k-mers (sketches) of every file,
you can also use <code>--by-seq</code> to compute for every sequence,
where sequence IDs in all input files are better to be distinct.
It also <strong>supports splitting sequences into chunks, this
could increase the specificity in profiling results at the cost
of a slower searching speed</strong>. </p>
<p><strong>Splitting sequences</strong>:</p>
<ol>
<li>Sequences can be split into chunks by a chunk size 
    (<code>-s/--split-size</code>) or number of chunks (<code>-n/--split-number</code>)
    with overlap (<code>-l/--split-overlap</code>).
    <strong><em>In this mode, the sequences of each genome should be saved in an
    individual file</em></strong>.</li>
<li>When splitting by number of chunks, <strong>all sequences (except for
    these matching any regular expression given by <code>-B/--seq-name-filter</code>)
    in a sequence file are concatenated with k-1 Ns before splitting</strong>.</li>
<li>Both sequence/reference IDs and chunks indices are saved for later use,
    in form of meta/description data in <code>.unik</code> files, and will
    be reported in <code>kmcp search</code> results.</li>
</ol>
<p><strong>Metadata</strong>:</p>
<ol>
<li>Every outputted <code>.unik</code> file contains the sequence/reference ID,
    chunk index, number of chunks, and genome size of reference.</li>
<li>When parsing whole sequence files or splitting by number of chunks,
    the <strong>identifier of a reference</strong> is the basename of the input file
    by default. It can also be extracted from the input file name via
    <code>-N/--ref-name-regexp</code>, e.g., <code>^(\w{3}_\d{9}\.\d+)</code> for <strong>RefSeq assembly accessions</strong>.</li>
</ol>
<p>Multiple <strong>sizes of k-mers</strong> are supported, but a single k-mer size
is good enough. </p>
<p><strong>Supported k-mer (sketches) types</strong>:</p>
<ol>
<li>K-mer:<ul>
<li>ntHash of k-mer (<code>-k</code>)</li>
</ul>
</li>
<li>K-mer sketchs (all using ntHash):<ul>
<li>Scaled MinHash (<code>-k -D</code>), previously named Scaled MinHash</li>
<li>Minimizer      (<code>-k -W</code>), optionally scaling/down-sampling (<code>-D</code>)</li>
<li>Closed Syncmer (<code>-k -S</code>), optionally scaling/down-sampling (<code>-D</code>)</li>
</ul>
</li>
</ol>
<p><strong>Output</strong>:</p>
<ol>
<li>
<p>All outputted <code>.unik</code> files are saved in <code>${outdir}</code>, with path</p>
<pre><code>${outdir}/xxx/yyy/zzz/${infile}-id_${seqID}.unik
</code></pre>
<p>where dirctory tree <code>/xxx/yyy/zzz/</code> is built for &gt; 1000 output files.</p>
</li>
<li>
<p>For splitting sequence mode (<code>--split-size &gt; 0</code> or <code>--split-number &gt; 0</code>),
    output files are:</p>
<pre><code>${outdir}//xxx/yyy/zzz/${infile}/{seqID}-chunk_${chunkIdx}.unik
</code></pre>
</li>
<li>
<p>A summary file (<code>${outdir}/_info.txt</code>) is generated for later use.
     <strong><em>Users need to check if the reference IDs (column <code>name</code>) are what
     supposed to be</em></strong>.</p>
</li>
</ol>
<p><strong>Performance tips</strong>:</p>
<ol>
<li>Decrease value of <code>-j/--threads</code> for data in hard disk drives (HDD) to
    reduce I/O pressure.</li>
</ol>
<p><strong>Commands</strong> (<a href="/kmcp/usage/#compute">usage</a>):</p>
<pre><code># compute k-mers
#   sequence containing "plasmid" in name are ignored,
#   reference genomes are split into 10 chunks,
#   k = 21
kmcp compute --in-dir refs/ \
    --kmer 21 \
    --split-number 10 \
    --split-overlap 150 \
    --seq-name-filter plasmid \
    --ref-name-regexp '(.+).fasta.gz' \
    --out-dir refs-k21-n10

# demo output
13:11:13.397 [INFO] kmcp v0.9.0
13:11:13.397 [INFO]   https://github.com/shenwei356/kmcp
13:11:13.397 [INFO] 
13:11:13.397 [INFO] checking input files ...
13:11:13.398 [INFO]   9 input file(s) given
13:11:13.398 [INFO] 
13:11:13.398 [INFO] -------------------- [main parameters] --------------------
13:11:13.398 [INFO] input and output:
13:11:13.398 [INFO]   input directory: refs/
13:11:13.398 [INFO]     regular expression of input files: (?i)\.(f[aq](st[aq])?|fna)(.gz)?$
13:11:13.398 [INFO]     *regular expression for extracting reference name from file name: (?i)(.+).fasta.gz
13:11:13.398 [INFO]     *regular expressions for filtering out sequences: [plasmid]
13:11:13.398 [INFO]   output directory: refs-k21-n10
13:11:13.398 [INFO] 
13:11:13.398 [INFO] sequences splitting: true
13:11:13.398 [INFO]   split parts: 10, overlap: 100 bp
13:11:13.398 [INFO] 
13:11:13.398 [INFO] k-mer (sketches) computing:
13:11:13.398 [INFO]   k-mer size(s): 21
13:11:13.398 [INFO]   circular genome: false
13:11:13.398 [INFO]   saving exact number of k-mers: true
13:11:13.398 [INFO] 
13:11:13.398 [INFO] -------------------- [main parameters] --------------------
13:11:13.398 [INFO] 
13:11:13.398 [INFO] computing ...
processed files:  9 / 9 [======================================] ETA: 0s. done
13:11:13.845 [INFO] 
13:11:13.845 [INFO] elapsed time: 453.870411ms
13:11:13.845 [INFO]
</code></pre>
<p>A summary file (<code>_info.txt</code>) is generated for later use.
<strong><em>Users need to check if the reference IDs (column <code>name</code>) are what supposed to be</em></strong>.</p>
<table>
<thead>
<tr>
<th align="left">#path</th>
<th align="left">name</th>
<th align="left">chunkIdx</th>
<th align="left">idxNum</th>
<th align="left">genomeSize</th>
<th align="left">kmers</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">refs-k21-n10/672/060/672/NC_010655.1.fasta.gz/NC_010655.1-chunk_0.unik</td>
<td align="left">NC_010655.1</td>
<td align="left">0</td>
<td align="left">10</td>
<td align="left">2664102</td>
<td align="left">264247</td>
</tr>
<tr>
<td align="left">refs-k21-n10/672/060/672/NC_010655.1.fasta.gz/NC_010655.1-chunk_1.unik</td>
<td align="left">NC_010655.1</td>
<td align="left">1</td>
<td align="left">10</td>
<td align="left">2664102</td>
<td align="left">266237</td>
</tr>
<tr>
<td align="left">refs-k21-n10/595/162/595/NC_012971.2.fasta.gz/NC_012971.2-chunk_0.unik</td>
<td align="left">NC_012971.2</td>
<td align="left">0</td>
<td align="left">10</td>
<td align="left">4558953</td>
<td align="left">450494</td>
</tr>
<tr>
<td align="left">refs-k21-n10/292/039/292/NC_000913.3.fasta.gz/NC_000913.3-chunk_0.unik</td>
<td align="left">NC_000913.3</td>
<td align="left">0</td>
<td align="left">10</td>
<td align="left">4641652</td>
<td align="left">459277</td>
</tr>
<tr>
<td align="left">refs-k21-n10/934/859/934/NC_013654.1.fasta.gz/NC_013654.1-chunk_0.unik</td>
<td align="left">NC_013654.1</td>
<td align="left">0</td>
<td align="left">10</td>
<td align="left">4717338</td>
<td align="left">470575</td>
</tr>
</tbody>
</table>
<p>Meta data in the <code>.unik</code> file can be showed using <code>kmcp utils unik-info</code>:</p>
<pre><code>kmcp utils unik-info refs-k21-n10/072/380/072/NZ_CP028116.1.fasta.gz/NZ_CP028116.1-chunk_0.unik -a
</code></pre>
<h3 id="step-2-building-databases">Step 2. Building databases</h3>
<p>KMCP builds index for k-mers (sketches) with a modified Compact Bit-sliced
Signature Index (<a href="https://arxiv.org/abs/1905.09624">COBS</a>). 
We completely rewrite the algorithms, data structure, and file format,
and have improved the indexing and searching speed
(check <a href="/kmcp/benchmark/searching">benchmark</a>).</p>
<p><strong>Input</strong>:</p>
<ul>
<li>The output directory generated by <code>kmcp compute</code>.</li>
</ul>
<p><strong>Database size and searching accuracy</strong>:</p>
<ol>
<li><strong>Use <code>--dry-run</code> to adjust parameters and check the final number of 
    index files (#index-files) and the total file size</strong>.</li>
<li><code>-f/--false-positive-rate</code>: the default value <code>0.3</code> is enough for a
    query with tens of k-mers (see BIGSI/COBS paper).
    Small values could largely increase the size of the database.</li>
<li><code>-n/--num-hash</code>: large values could reduce the database size,
    at the cost of a slower searching speed. Values &lt;=4 are recommended.</li>
<li>
<p>The value of block size <code>-b/--block-size</code> is better to be multiple of 64.
    The default value is: </p>
<pre><code>(#unikFiles/#threads + 7) / 8 * 8
</code></pre>
</li>
<li>
<p>Use flag <code>-x/--block-sizeX-kmers-t</code>, <code>-8/--block-size8-kmers-t</code>,
    and <code>-1/--block-size1-kmers-t</code> to separately create indexes for
    inputs with huge number of k-mers, for precise control of
    database size.</p>
</li>
</ol>
<p><strong>Taxonomy data</strong>:</p>
<ol>
<li>No taxonomy data are included in the database.</li>
<li>Taxonomy information are only needed in <code>kmcp profile</code>.</li>
</ol>
<p><strong>Performance tips</strong>:</p>
<ol>
<li>The umber of blocks (<code>.uniki</code> files) is better to be smaller than or equal
   to the number of CPU cores for faster searching speed. 
   <strong>We can set <code>-j/--threads</code> to control the blocks number</strong>.
   When more threads (&gt;= 1.3 * #blocks) are given, extra workers are
   automatically created.</li>
<li><code>#threads</code> files are simultaneously opened, and the max number
    of opened files is limited by the flag <code>-F/--max-open-files</code>.
    You may use a small value of <code>-F/--max-open-files</code> for 
    hard disk drive storages.</li>
<li>When the database is used in a new computer with more CPU cores,
   <code>kmcp search</code> could automatically scale to utilize as many cores
   as possible.</li>
</ol>
<p><strong>Commands</strong> (<a href="/kmcp/usage/#compute">usage</a>):</p>
<pre><code># build database
#   number of index files: 32, for server with &gt;= 32 CPU cores
#   bloom filter parameter:
#     number of hash function: 1
#     false positive rate: 0.3
kmcp index -I refs-k21-n10 \
    --threads 32 \
    --num-hash 1 \
    --false-positive-rate 0.3 \
    --out-dir refs.kmcp

# demo output
22:44:17.710 [INFO] kmcp v0.9.0
22:44:17.710 [INFO]   https://github.com/shenwei356/kmcp
22:44:17.710 [INFO] 
22:44:17.710 [INFO] loading .unik file infos from file: refs-k21-n10/_info.txt
22:44:17.716 [INFO]   90 cached file infos loaded
22:44:17.716 [INFO] 
22:44:17.716 [INFO] -------------------- [main parameters] --------------------
22:44:17.716 [INFO]   number of hashes: 1
22:44:17.716 [INFO]   false positive rate: 0.300000
22:44:17.717 [INFO]   k-mer size(s): 21
22:44:17.717 [INFO]   split seqequence size: 0, overlap: 0
22:44:17.717 [INFO]   block-sizeX-kmers-t: 10.00 M
22:44:17.717 [INFO]   block-sizeX        : 256.00
22:44:17.717 [INFO]   block-size8-kmers-t: 20.00 M
22:44:17.717 [INFO]   block-size1-kmers-t: 200.00 M
22:44:17.717 [INFO] -------------------- [main parameters] --------------------
22:44:17.717 [INFO] 
22:44:17.717 [INFO] building index ...
22:44:17.726 [WARN] ignore -X/--block-size (256) which is &gt;= -b/--block-size (8)
22:44:17.726 [INFO] 
22:44:17.726 [INFO]   block size: 8
22:44:17.726 [INFO]   number of index files: 32 (may be more)
22:44:17.726 [INFO] 
22:44:17.726 [block #001]   1 / 1  100 %
22:44:17.726 [block #002]   1 / 1  100 %
22:44:17.726 [block #003]   1 / 1  100 %
22:44:17.726 [block #004]   1 / 1  100 %
22:44:17.726 [block #005]   1 / 1  100 %
22:44:17.726 [block #006]   1 / 1  100 %
22:44:17.726 [block #007]   1 / 1  100 %
22:44:17.726 [block #008]   1 / 1  100 %
22:44:17.726 [block #009]   1 / 1  100 %
22:44:17.727 [block #010]   1 / 1  100 %
22:44:17.727 [block #011]   1 / 1  100 %
22:44:17.727 [block #012]   1 / 1  100 %
[saved index files] 12 / 12  ETA: 0s. done
22:44:17.933 [INFO] 
22:44:17.933 [INFO] kmcp database with 42713316 k-mers saved to refs.kmcp
22:44:17.933 [INFO] total file size: 15.66 MB
22:44:17.933 [INFO] total index files: 12
22:44:17.933 [INFO] 
22:44:17.933 [INFO] elapsed time: 223.524128ms
22:44:17.933 [INFO]
</code></pre>
<p>Output:</p>
<pre><code>refs.kmcp/
 R001
     _block001.uniki
     _block002.uniki
     _block003.uniki
     _block004.uniki
     _block005.uniki
     _block006.uniki
     _block007.uniki
     _block008.uniki
     _block009.uniki
     _block010.uniki
     _block011.uniki
     _block012.uniki
     __db.yml
     __name_mapping.tsv
</code></pre>
<p><code>__db.yml</code> contains configuration of the database, and <code>_blockXXX.uniki</code> are index files.
<code>kmcp utils index-info</code> could show the basic information of index files:</p>
<pre><code>kmcp utils index-info  refs.kmcp/R001/_block001.uniki
</code></pre>
<table>
<thead>
<tr>
<th align="left">file</th>
<th align="left">k</th>
<th align="left">canonical</th>
<th align="left">num-hashes</th>
<th align="left">num-sigs</th>
<th align="left">num-names</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">refs.kmcp/R001/_block001.uniki</td>
<td align="left">21</td>
<td align="left">true</td>
<td align="left">1</td>
<td align="left">746442</td>
<td align="left">8</td>
</tr>
</tbody>
</table>
<p>Where <code>num-sigs</code> is the size of the bloom filters, and <code>num-names</code> is the number of genome (chunks).</p>
<p>What's next? Check the <a href="/kmcp/tutorial">tutorials</a>.</p>
<h2 id="merging-gtdb-and-ncbi-taxonomy">Merging GTDB and NCBI taxonomy</h2>
<p>By default, we use NCBI taxonomy database for reference genomes from GTDB and Refseq.
Alternately, we can also use the GTDB taxonomy.</p>
<p>The idea is to export lineages from both GTDB and NCBI using <a href="https://bioinf.shenwei.me/taxonkit/usage/#reformat">taxonkit reformat</a>, and then create taxdump files from them with <a href="https://bioinf.shenwei.me/taxonkit/usage/#create-taxdump">taxonkit create-taxdump</a>.</p>
<ol>
<li>
<p>Mapping genome assembly accessions to GTDB lineages using existing <a href="https://github.com/shenwei356/gtdb-taxdump">GTDB-taxdump</a>,
   which provids GTDB taxonomy taxdump files for each GTDB version and a <code>taxid.map</code> file mapping accessions to TaxIds of taxa at the subspecies rank.
   Here, we only output lineages down to the species rank and leave name of taxa at the subspecies/strain rank empty.</p>
<pre><code>taxonkit reformat \
    --data-dir gtdb-taxdump/R214/ \
    --taxid-field 2 \
    --format "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t" \
    gtdb-taxdump/R214/taxid.map \
| csvtk cut -H -t -f -2 -o gtdb-r214.tsv
</code></pre>
</li>
<li>
<p>Exporting taxonomic lineages of viral and fungal taxa with rank equal to or lower than species from NCBI taxdump.
   For taxa whose rank is "no rank" below the species, we treat them as tax of strain rank (<code>--pseudo-strain</code>, taxonkit v0.14.1 or later versionsneeded).</p>
<pre><code># for viral data
taxonkit reformat \
    --data-dir taxdump/ \
    --taxid-field 2 \
    --pseudo-strain \
    --format "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}" \
    taxid.map \
| csvtk cut -H -t -f -2 -o genbank-viral.tsv

# for fungal data
taxonkit reformat \
    --data-dir taxdump/ \
    --taxid-field 2 \
    --pseudo-strain \
    --format "{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}\t{t}" \
    taxid.map \
| csvtk cut -H -t -f -2 -o refseq-fungi.tsv
</code></pre>
</li>
<li>
<p>Creating taxdump from lineages above.</p>
<pre><code># move the lineage files to current directory
# cp gtdb/gtdb-r214.tsv genbank-viral/genbank-viral.tsv refseq-fungi/refseq-fungi.tsv .

cat gtdb-r214.tsv genbank-viral.tsv refseq-fungi.tsv \
    | taxonkit create-taxdump \
        --field-accession 1 \
        -R "superkingdom,phylum,class,order,family,genus,species,strain" \
        -O taxdump.gtdb+ncbi
</code></pre>
</li>
<li>
<p>NCBI taxonomy of Viruses changed rapidly, of which some TaxIds might be deleted.
   You may added them to the the <code>taxid.map</code> file for compatibility.</p>
<pre><code>cat genbank-viral.tsv refseq-fungi.tsv gtdb-r214.tsv \
    | csvtk grep -Ht -f 2 -r -p "^$" \
    | cut -f 1 \
    | awk '{print $1"\t0"}' \
    &gt;&gt; taxdump.gtdb+ncbi/taxid.map
</code></pre>
</li>
</ol>
<p>Some tests:</p>
<pre><code># SARS-COV-2 in NCBI taxonomy
$ echo 2697049 \
    | taxonkit lineage -t --data-dir taxdump/ \
    | csvtk cut -Ht -f 3 \
    | csvtk unfold -Ht -f 1 -s ";" \
    | taxonkit lineage -r -n -L --data-dir taxdump/ \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk pretty -Ht
10239     superkingdom   Viruses
2559587   clade          Riboviria
2732396   kingdom        Orthornavirae
2732408   phylum         Pisuviricota
2732506   class          Pisoniviricetes
76804     order          Nidovirales
2499399   suborder       Cornidovirineae
11118     family         Coronaviridae
2501931   subfamily      Orthocoronavirinae
694002    genus          Betacoronavirus
2509511   subgenus       Sarbecovirus
694009    species        Severe acute respiratory syndrome-related coronavirus
2697049   no rank        Severe acute respiratory syndrome coronavirus 2

$ echo "Severe acute respiratory syndrome coronavirus 2" | taxonkit name2taxid --data-dir taxdump.gtdb+ncbi/
Severe acute respiratory syndrome coronavirus 2 216305222

$ echo 216305222 \
    | taxonkit lineage -t --data-dir taxdump.gtdb+ncbi/ \
    | csvtk cut -Ht -f 3 \
    | csvtk unfold -Ht -f 1 -s ";" \
    | taxonkit lineage -r -n -L --data-dir taxdump.gtdb+ncbi/ \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk pretty -Ht
1287770734   superkingdom   Viruses
1506901452   phylum         Pisuviricota
1091693597   class          Pisoniviricetes
37745009     order          Nidovirales
738421640    family         Coronaviridae
906833049    genus          Betacoronavirus
1015862491   species        Severe acute respiratory syndrome-related coronavirus
216305222    strain         Severe acute respiratory syndrome coronavirus 2



$ echo "Escherichia coli"  | taxonkit name2taxid --data-dir taxdump.gtdb+ncbi/
Escherichia coli        1945799576

$ echo 1945799576 \
    | taxonkit lineage -t --data-dir taxdump.gtdb+ncbi/ \
    | csvtk cut -Ht -f 3 \
    | csvtk unfold -Ht -f 1 -s ";" \
    | taxonkit lineage -r -n -L --data-dir taxdump.gtdb+ncbi/ \
    | csvtk cut -Ht -f 1,3,2 \
    | csvtk pretty -Ht
609216830    superkingdom   Bacteria
1641076285   phylum         Proteobacteria
329474883    class          Gammaproteobacteria
1012954932   order          Enterobacterales
87250111     family         Enterobacteriaceae
1187493883   genus          Escherichia
1945799576   species        Escherichia coli
</code></pre>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../download/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Download" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Download
            </div>
          </div>
        </a>
      
      
        
        <a href="../usage/" class="md-footer__link md-footer__link--next" aria-label="Next: Usage" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Usage
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../assets/javascripts/workers/search.b6ff8c07.min.js"}</script>
    
    
      <script src="../assets/javascripts/bundle.5e13d264.min.js"></script>
      
    
  </body>
</html>