
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="KMCP - K-mer-based Metagenomic Classification and Profiling">
      
      
        <meta name="author" content="Wei Shen">
      
      
        <link rel="canonical" href="https://bioinf.shenwei.me/kmcp/tutorial/detecting-contaminated-seqs/">
      
      <link rel="icon" href="../../favicon.svg">
      <meta name="generator" content="mkdocs-1.2.3, mkdocs-material-8.1.5">
    
    
      
        <title>Detecting contaminated sequences - KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.ee0f47ba.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.e6a45f82.min.css">
        
          
          
          <meta name="theme-color" content="#009485">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="teal" data-md-color-accent="blue-grey">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#detecting-contaminated-sequences-in-a-bacteria-assembly" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching" class="md-header__button md-logo" aria-label="KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching" data-md-component="logo">
      
  <img src="../../favicon.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Detecting contaminated sequences
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/shenwei356/kmcp/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching" class="md-nav__button md-logo" aria-label="KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching" data-md-component="logo">
      
  <img src="../../favicon.svg" alt="logo">

    </a>
    KMCP - accurate metagenomic profiling and fast large-scale sequence/genome searching
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/shenwei356/kmcp/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../download/" class="md-nav__link">
        Download
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../database/" class="md-nav__link">
        Databases
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../usage/" class="md-nav__link">
        Usage
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          Tutorials
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Tutorials" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Tutorials
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../profiling/" class="md-nav__link">
        Taxonomic profiling
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../detecting-pathogens/" class="md-nav__link">
        Detecting specific pathogens
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Detecting contaminated sequences
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Detecting contaminated sequences
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#changelog" class="md-nav__link">
    Changelog
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tools" class="md-nav__link">
    Tools
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#database" class="md-nav__link">
    Database
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#hardware" class="md-nav__link">
    Hardware
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steps" class="md-nav__link">
    Steps
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../searching/" class="md-nav__link">
        Sequence and genome searching
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          Benchmarks
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Benchmarks" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          Benchmarks
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../benchmark/profiling/" class="md-nav__link">
        Taxonomic profiling
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../benchmark/searching/" class="md-nav__link">
        Sequence and genome searching
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../faq/" class="md-nav__link">
        FAQs
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356" class="md-nav__link">
        More tools
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#changelog" class="md-nav__link">
    Changelog
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tools" class="md-nav__link">
    Tools
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#database" class="md-nav__link">
    Database
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#hardware" class="md-nav__link">
    Hardware
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#steps" class="md-nav__link">
    Steps
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
<a href="https://github.com/shenwei356/kmcp/edit/master/docs/tutorial/detecting-contaminated-seqs/index.md" title="Edit this page" class="md-content__button md-icon">
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
</a>


<h1 id="detecting-contaminated-sequences-in-a-bacteria-assembly">Detecting contaminated sequences in a bacteria assembly</h1>
<h2 id="changelog">Changelog</h2>
<p>Tracked changes in <a href="https://github.com/shenwei356/kmcp/commits/main/docs/tutorial/detecting-contaminated-seqs/index.md">github</a></p>
<ul>
<li>2023-07-20<ul>
<li>Recommend using masked GTDB index files</li>
<li>Recommend using GTDB taxdump files rather than NCBI taxdump files.</li>
</ul>
</li>
<li>2023-07-16<ul>
<li>First public version.</li>
</ul>
</li>
</ul>
<h2 id="tools">Tools</h2>
<ul>
<li>kmcp: https://github.com/shenwei356/kmcp</li>
<li>seqkit: &gt;= v2.5.0 which has the new command <code>seqkit merge-slides</code>.</li>
<li>taxonkit: https://github.com/shenwei356/taxonkit</li>
<li>csvtk: https://github.com/shenwei356/csvtk</li>
</ul>
<h2 id="database">Database</h2>
<ul>
<li>KMCP database v2023.05: please choose files below to download from <a href="https://1drv.ms/f/s!Ag89cZ8NYcqtlhYlyHuaKweYNu93?e=9NOxqg">here</a>.<ul>
<li>GTDB index files (need to be uncompressed)<ul>
<li>Option 1: <strong>(Recommended) with both plasmid and prophage masked</strong> with <a href="https://github.com/apcamargo/genomad">genomad</a>.<ul>
<li><code>gtdb_masked.part_1.kmcp.tar.gz</code></li>
<li><code>gtdb_masked.part_2.kmcp.tar.gz</code></li>
</ul>
</li>
<li>Option 2: <strong>only plasmid filtered out</strong> according to sequence names.<ul>
<li><code>gtdb.part_1.kmcp.tar.gz</code></li>
<li><code>gtdb.part_2.kmcp.tar.gz</code></li>
</ul>
</li>
</ul>
</li>
<li>Taxdump files and TaxId mapping file:<ul>
<li>Option 1: <strong>(Recommended) GTDB+NCBI taxonomy</strong>. Actually, only the GTDB taxonomy is used here, cause we only use the GTDB KMCP database, no viral or fungal databases are used.<ul>
<li><code>taxdump.gtdbR214.1+ncbi.tar.gz</code>, uncompressed directory is <code>taxdump.gtdb+ncbi/</code></li>
<li><code>taxid.gtdbR214.1+ncbi.map</code></li>
</ul>
</li>
<li>Option 2: <strong>NCBI taxonomy</strong><ul>
<li><code>taxdump.tar.gz</code>, uncompressed directory is <code>taxdump/</code></li>
<li><code>taxid.map</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<h2 id="hardware">Hardware</h2>
<ul>
<li>RAM &gt;= 64GB</li>
<li>CPUs &gt;= 32 preferred.</li>
</ul>
<h2 id="steps">Steps</h2>
<p>Generating reads (sliding a 200-bp windows with a step of 50-bp) from the genome,
and performing metagenomoic profiling with them.</p>
<pre><code># variables
input=SAMN02360712.contigs.fa.gz
taxdump=~/ws/data/kmcp2023/taxdump.gtdb+ncbi/
taxmap=~/ws/data/kmcp2023/taxid.gtdbR214.1+ncbi.map

# search against GTDB databases
# !!! if the KMCP databases are in a network-attached storage disk (NAS),
# !!! please add the flag "-w" to "kmcp search"
for db in ~/ws/data/kmcp2023/gtdb_masked.part_{1,2}.kmcp/; do
    seqkit sliding -g -s 50 -W 200 $input \
        | kmcp search -w -d $db -o $input.kmcp@$(basename $db).tsv.gz
done

# merge seach results
kmcp merge -o $input.kmcp.tsv.gz $input.kmcp@*.tsv.gz

# profiling and outputing read classification results.
# here the profiling mode 0 with the highest sensitivity is used.
kmcp profile -X $taxdump -T $taxmap $input.kmcp.tsv.gz -m 0 \
    -o $input.kmcp.tsv.gz.k.profile -B $input.kmcp.tsv.gz.binning.gz

cat $input.kmcp.tsv.gz.k.profile \
    | csvtk cut -t -f ref,percentage,score,chunksFrac,reads,taxid,taxname \
    | csvtk pretty -t

ref               percentage   score    chunksFrac   reads    taxid        taxname
---------------   ----------   ------   ----------   ------   ----------   -------------------------
GCF_001457615.1   99.579056    100.00   1.00         113704   1696268695   Pseudomonas aeruginosa
GCF_000008985.1   0.062945     100.00   1.00         23       418881019    Francisella tularensis
GCF_000017205.1   0.289416     90.83    1.00         345      1859336444   Pseudomonas aeruginosa_A
GCF_002091755.1   0.037182     83.61    0.40         39       744524063    Pseudomonas nitroreducens
GCF_003797765.1   0.012409     85.83    0.20         14       49646394     Ramlibacter sp003797765
GCF_009735585.1   0.006522     77.50    0.20         4        182564388    Gramella aestuarii
GCF_900112375.1   0.012470     66.94    0.20         13       132541362    Pseudomonas citronellolis
</code></pre>
<p>Checking contaminated regions. Here we check "inter-genus" contamination.</p>
<pre><code># the taxid of the predominant target of the highest percentage
taxidMain=$(csvtk cut -t -f taxid $input.kmcp.tsv.gz.k.profile | sed 1d | head -n 1)

# the genus taxid of the target.
# the placeholders of different ranks:
#    {k}: superkingdom
#    {K}: kingdom
#    {p}: phylum
#    {c}: class
#    {o}: order
#    {f}: family
#    {g}: genus
#    {s}: species
taxidMainGenus=$(echo $taxidMain | taxonkit reformat --data-dir $taxdump -I 1 -t -f '{g}' | cut -f 3)

# find out sequence IDs belonging to other genera (potential contaminated contigs).
#   1. filter out reads with taxids equal to and below the genus node of the predominant target.
#   2. filter out reads with taxids above the genus node of the predominant target.
zcat $input.kmcp.tsv.gz.binning.gz | grep -v ^@ \
    | csvtk grep -Ht -f 2 -v -P &lt;(taxonkit list --data-dir $taxdump --ids $taxidMainGenus -I "") \
    | csvtk grep -Ht -f 2 -v -P &lt;(csvtk cut -t -f taxpathsn $input.kmcp.tsv.gz.k.profile \
                                    | sed 1d | head -n 1 | sed "s/;/\n/g") \
        -o $input.kmcp.tsv.gz.binning.filtered.tsv

# merge regions. seqkit &gt;=v2.5.0 is needed.
seqkit merge-slides $input.kmcp.tsv.gz.binning.filtered.tsv --quiet \
    -o $input.kmcp.tsv.gz.cont.tsv

# collect taxonomy info for each contig
csvtk mutate -Ht -p '^(.+)_sliding' $input.kmcp.tsv.gz.binning.filtered.tsv \
    | csvtk freq -Ht -f 3,2 -nr \
    | taxonkit lineage --data-dir $taxdump -i 2 -n -L \
    | awk -F'\t' '{print $1"\t"$2"("$4")"}' \
    | csvtk fold -Ht -f 1 -v 2 -s ',' \
        -o $input.kmcp.tsv.gz.binning.filtered.tsv.taxa

# append contig length and taxid(s)
csvtk join -Ht $input.kmcp.tsv.gz.cont.tsv &lt;(seqkit fx2tab -ni -l $input) \
    | awk '{print $0"\t"($3-$2)"\t"($3-$2)/$4}' \
    | csvtk join -Ht - $input.kmcp.tsv.gz.binning.filtered.tsv.taxa \
    | csvtk add-header -Ht -n chr,begin,end,contig_len,len,proportion,taxa \
    | csvtk sort -t -k proportion:nr \
    | tee $input.kmcp.tsv.gz.cont.details.tsv \
    | csvtk pretty -t

chr                        begin    end      contig_len   len    proportion    taxa
------------------------   ------   ------   ----------   ----   -----------   ---------------------------------------------------------------
SAMN02360712.contig00044   0        1151     1151         1151   1             418881019(Francisella tularensis)
SAMN02360712.contig00028   26600    26800    47311        200    0.00422735    49646394(Ramlibacter sp003797765)
SAMN02360712.contig00012   163700   163900   164357       200    0.00121686    182564388(Gramella aestuarii),49646394(Ramlibacter sp003797765)
SAMN02360712.contig00012   155750   155950   164357       200    0.00121686    182564388(Gramella aestuarii),49646394(Ramlibacter sp003797765)
SAMN02360712.contig00001   454500   455250   670989       750    0.00111775    49646394(Ramlibacter sp003797765)
SAMN02360712.contig00002   362200   362500   622965       300    0.000481568   182564388(Gramella aestuarii)
</code></pre>
<p>We can see the whole (proportion: 1) contig <code>SAMN02360712.contig00044</code> is from a species belonging to a different genus even a different order (see the table below). So it is likely be a contaminated sequence.</p>
<pre><code>echo 1696268695 418881019 \
    | sed -E 's/\s+/\n/g' \
    | taxonkit reformat --data-dir $taxdump -I 1 -f '{k}\t{p}\t{c}\t{o}\t{f}\t{g}\t{s}' \
    | csvtk add-header -t -n taxid,kingdom,phylum,class,order,family,genus,species \
    | csvtk csv2md -t
</code></pre>
<table>
<thead>
<tr>
<th align="left">taxid</th>
<th align="left">kingdom</th>
<th align="left">phylum</th>
<th align="left">class</th>
<th align="left">order</th>
<th align="left">family</th>
<th align="left">genus</th>
<th align="left">species</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">1696268695</td>
<td align="left">Bacteria</td>
<td align="left">Pseudomonadota</td>
<td align="left">Gammaproteobacteria</td>
<td align="left">Pseudomonadales</td>
<td align="left">Pseudomonadaceae</td>
<td align="left">Pseudomonas</td>
<td align="left">Pseudomonas aeruginosa</td>
</tr>
<tr>
<td align="left">418881019</td>
<td align="left">Bacteria</td>
<td align="left">Pseudomonadota</td>
<td align="left">Gammaproteobacteria</td>
<td align="left">Francisellales</td>
<td align="left">Francisellaceae</td>
<td align="left">Francisella</td>
<td align="left">Francisella tularensis</td>
</tr>
</tbody>
</table>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../detecting-pathogens/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Detecting specific pathogens" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Detecting specific pathogens
            </div>
          </div>
        </a>
      
      
        
        <a href="../searching/" class="md-footer__link md-footer__link--next" aria-label="Next: Sequence and genome searching" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Sequence and genome searching
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../assets/javascripts/workers/search.b6ff8c07.min.js"}</script>
    
    
      <script src="../../assets/javascripts/bundle.5e13d264.min.js"></script>
      
    
  </body>
</html>